### ISSUE TITLE
Conditional/Inactive hyperparameters

### ISSUE THREAD

**Issue by sebastianpinedaar** [awslabs/syne-tune#743]
Hi, 

I am using a training script with conditional hyperparameters, which might be "None" as input arguments. However, I do not know how I could deal with it. If there is no direct way, I think it would be a nice feature.

Specifically, assume I have the following training script:

```python
# train_height_simple.py
import logging
import time

from syne_tune import Reporter
from argparse import ArgumentParser

if __name__ == '__main__':
    root = logging.getLogger()
    root.setLevel(logging.INFO)
    parser = ArgumentParser()
    parser.add_argument('--epochs', type=int)
    parser.add_argument('--width', type=float)
    parser.add_argument('--height', type=float)

    #problematic argument
    parser.add_argument('--depth', action='store_true', default=None)

    parser.
    args, _ = parser.parse_known_args()
    report = Reporter()
    for step in range(args.epochs):
        time.sleep(0.1)
        dummy_score = 1.0 / (0.1 + args.width * step / 100) + args.height * 0.1
        # Feed the score back to Syne Tune.
        report(epoch=step + 1, mean_loss=dummy_score)
```

Assume that the possible values of depth are a list $x \in (-1,1)$, but it can also be none (when the depth is inactive).  Then the question is: how should I define the config space for `depth`?

```python
# hyperparameter search space to consider
config_space = {
    'width': randint(1, 20),
    'height': randint(1, 20),
    'depth': ?
    'epochs': 100,
}
```
Is there a way to deal with it that I may be missing? If not, maybe a new type of class in `syne_tune.config_space` could be necessary.

Another related use case is when dealing with conditional hyperparameters. Specifically, I realized that when training a neural network using the `timm` library script [here](https://github.com/huggingface/pytorch-image-models/blob/main/train.py). The problem arises, for instance, when I want to use a config space like this:

```python
# hyperparameter search space to consider
config_space = {
   optimizer : choice(["adam", "sgd"]),
   opt_betas : choice(["0.9 0.99", "0.99"])
}
```
In this case, `opt_betas` is just active when the optimizer is `adam`. If the optimizer is `sgd`, then it should be `None`. Otherwise, the script fails. How could we enforce this condition for the optimizer?

Thanks,

Sebastian

**Comment by geoalgo** [comment]
Hi Sebastian,

Syne Tune does not support conditional hyperparameters natively yet. 

I see several several options:

1. you can write your own `TrialSheduler` (see this [example](https://github.com/awslabs/syne-tune/blob/main/examples/launch_height_standalone_scheduler.py)). There you can write your own suggestion function and apply whatever logic you want (you could then use `ConfigSpace` for instance which supports hierarchical spaces).

2. You can override a scheduler and change `_postprocess_config` to set to None some hyperparameters conditioned on the sampled value.

3. Add support for those type of sampling in `config_space` as you suggested.

The downside of 1. is that you will have to rewrite your optimizer however, this is something you likely would have to do as our BO does not support such spaces for instance. The upside is that you have all flexibility to add your method.

2. Is a bit in-between, you may be able to reuse some of the schedulers with this approach.

3. Is the cleanest but also an investment to be made to add clean support for conditional space.

Hope this helps, thanks for the issue!

In any case, let us know your thoughts and if you implement one of the approach we would be glad to receive your contribution as an example (easier) or an update to config-space (nice but probably longer and harder).

**Comment by mseeger** [comment]
What I mean:
```
   'depth': uniform(-1, 1),
   'depth_active': choice(["true", "false"]),
```

**Comment by mseeger** [comment]
For random search you get the same as if you encoded some complex conditional sampler for yourself (I'd not do that), and for BO, you may still outperform methods based on random search, even though the GP surrogate model does not model the conditional dependence of 'depth' on 'depth_active'.

**Comment by mseeger** [comment]
Of course, if you have a good idea how to support conditional relationships in a BO surrogate model, let us talk! There is some work, but this has no good general solution so far.
